README.md
目录
=================
- [目录](#目录)
- [概述](#概述)
- [端上SDK demo](#端上sdk-demo)

# 概述
Azero IoT Device SDK 是符合C99标准的嵌入式C程序库, 用于安全地接入Azero IoT平台, 仅需简单的设置即可以快速远程控制您的设备. 和AWS IoT 平台设备控制规则完全兼容, 目前支持Shadow服务.

# 端上SDK demo
```
#include "azero_iot_sdk_api.h"
#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <pthread.h>

static azero_iot_handler_t handler = NULL;
static void startiot();
/* 设备支持的控制策略, 和Azero IoT平台控制策略一致 */
static char *state_name[] = {"brightness","lightMode"};
static int   state_size = 2;

#define ProductID        "speaker_sai"         /* 一类设备的product Id */
#define DeviceSN         "AE:DE:C2:61:3F:18"   /* 本设备的SN, 本类设备中每个设备唯一, 可以为MAC或其他唯一标识 */
#define AccountPrefix    "32141iz1awewfq"      /* 一类设备的标识, IoT平台注册产生 */

// 断开连接后的重连处理
static bool reconnecting = false;
static void * _reconnectThread(void* args)
{
    pthread_detach(pthread_self());
    reconnecting = true;
    azero_iot_handler_release(handler);
    handler = NULL;
    startiot();
    sleep(2);
    reconnecting = false;
    return NULL;
}

//断开连接的回调
static void disconnect_cb(disconnect_reason_t reason)
{
    printf("disconnect reason %d\n",reason);
    if(reason == AZERO_IOT_MQTT_KEEP_ALIVE_TIMEOUT){
        if(false == reconnecting)
        {
            reconnecting = true;
            pthread_t ptid;
            pthread_create(&ptid, NULL, _reconnectThread,NULL);
        }
    }
}

static void deal_data_cb(const char *key, const char* value,size_t size, bool need_report)
{
    printf("get state= %s, value = %.*s,need report = %s\n",key,size,value,need_report?"true":"false");

    if(need_report){
        char *format = "{\"%s\":%.*s}";
        size_t length = strlen(format) + size + key;
        char data[length+1];
        sprintf(data,format,key,size,value);
        /* 端上状态上报, 如果是本地主动更改的状态, 最后一个参数传false, APP端发起的改变, 最后一个参数传true */
        azero_iot_shadow_update(handler,data,true);
    }
}

static void shadow_data_cb( shadow_data_type_t type, const char *pDocument, size_t documentLength)
{
    printf("shadow msg cb: %.*s\n",documentLength,pDocument);
    switch (type)
    {
    case AZERO_IOT_SHADOW_GET:
    case AZERO_IOT_SHADOW_DELTA:
        {
            bool stateFound = false, keyFound = false;
            const char * pState = NULL;
            size_t stateLength = 0;

            /* 查找json消息中是否有"state" key. */
            stateFound = azero_iot_parser_value( pDocument,
                                                    documentLength,
                                                    "state",
                                                    &pState,
                                                    &stateLength );
            if(!stateFound)
            {
                printf("pDocument is error\n");
                return;
            }
            /* 查找是否有已经定义好的策略 */
            printf("state = %.*s\n",stateLength,pState);
            for(int i=0; i<state_size; i++){
                
                const char * pData = NULL;
                size_t dataLength = 0;
                const char * keyName = state_name[i];
                keyFound = azero_iot_parser_value( pState,
                                                    stateLength,
                                                    keyName,
                                                    &pData,
                                                    &dataLength );
                printf(" find key = %s, result = %d\n",state_name[i],keyFound);
                if(keyFound)
                {
                    /* 对于get到的消息, 需要设备改变状态至相应的状态, 和IoT平台最新消息保持一致,
                    一般用于设备开机或者断网重连, 对APP端做出的改变进行同步;
                    delta 消息为APP下发控制指令后收到的消息, 需要将设备改变至相应状态, 同时将改变后状态上报IoT平台,
                    平台更新最新状态.
                    */
                    deal_data_cb(keyName,pData,dataLength,type==AZERO_IOT_SHADOW_DELTA);
                }
            }

            break;
        }
    case AZERO_IOT_SHADOW_DOCUMENT:
        {
            /* update后云端当前状态和上一个状态的通知, 可以依次查看是否上报错误 */
            break;
        }
    }
}

/* shadow 内容update结果回调, 成功或者失败*/
static void update_result_cb(const char* result)
{
    printf("update result is %s\n",result);
}

static void startiot()
{
    while(handler == NULL){
        /* global val, SDK will use this info later */
        static azero_iot_callback_t cbs = { .disconnect = disconnect_cb,\
                                            .param = shadow_data_cb,\
                                            .update_result = update_result_cb};
        static azero_device_info_t device_info = { .phostname = AccountPrefix".iot-dev.bj.soundai.cn",\ 
                                                    /* IoT 平台产生域名 */
                                                    .port = 8883,\
                                                    .accountprefix = AccountPrefix,\
                                                    .pIdentifier = AccountPrefix"-"DeviceSN,\
                                                    /* pIdentifier 每个设备唯一, 前缀必须为AccountPrefix */
                                                    .pThingName = ProductID""DeviceSN,\
                                                    /* pThingName 每个设备唯一, 必须为ProductIDDeviceSN拼接 */
                                                    /* IoT 平台产生以下key */
                                                    .inputCertPath = true,\
                                                    .pRootCa = "azero-api-cert.pem",\
                                                    .pClientCert = "1323214-cert.pem",\
                                                    .pPrivateKey = "1323214-private.key.pem"};
        device_info.pAttributeName = state_name;
        device_info.attributeSize = state_size;
        for(int i=0;i<state_size;i++)
        {
            printf("pAttributeName[%d] = %s\n",i,device_info.pAttributeName[i]);
        }

        handler = azero_iot_handler_create(&device_info,&cbs);
        sleep(2);        
    }
}

int main()
{
    startiot();

    azero_iot_shadow_get(handler);
    while(1){
        sleep(rand()%50+5);
        /* azero_iot_shadow_update(handler,"{\"brightness\":33}",false);*/
    };
    /*azero_iot_shadow_clear(handler);
    清空IoT 平台所有状态, 用于恢复出厂设置或者重置
    */
    /* 正常退出, 释放资源 */
    azero_iot_handler_release(handler);
    handler = NULL;
}

```

changelog.md
# v1.1.1
## fix publish payload size copy error
## update log file clear method
## change close network in ping timeout

# v1.1.0
## add freeRTOS ports

# v1.0.0
## add azero iot shadows service
## support linux64, arm32, aarch64 system

include/azero_iot_sdk_api.h
/**
 *  SoundAI Azero IOT SDK APIs
 *
 *  Copyright (C) 2020 SoundAI.com
 *  All Rights Reserved.
 */

#ifndef SOUNDAI_AZERO_IOT_API_H
#define SOUNDAI_AZERO_IOT_API_H

#ifndef SAI_API
    #if defined(_WIN32) && defined(DLL_EXPORT)
        #define SAI_API __declspec(dllexport)
    #elif defined(_WIN32)
        #define SAI_API __declspec(dllimport)
    #elif defined(__GNUC__) && defined(DLL_EXPORT)
        #define SAI_API __attribute__((visibility("default")))
    #else
        #define SAI_API
    #endif
#endif

#include <stdbool.h>
#include <stdlib.h>
#include <stdint.h>

#if defined(__cplusplus)
extern "C" 
{
#endif

/**
 * @ingroup mqtt_datatypes_enums
 * @brief Types of MQTT operations.
 *
 * The function @ref mqtt_function_operationtype can be used to get an operation
 * type's description.
 */
typedef enum azero_iot_mqtt_operation_type
{
    AZERO_IOT_MQTT_CONNECT,           /**< Client-to-server CONNECT. */
    AZERO_IOT_MQTT_PUBLISH_TO_SERVER, /**< Client-to-server PUBLISH. */
    AZERO_IOT_MQTT_PUBACK,            /**< Client-to-server PUBACK. */
    AZERO_IOT_MQTT_SUBSCRIBE,         /**< Client-to-server SUBSCRIBE. */
    AZERO_IOT_MQTT_UNSUBSCRIBE,       /**< Client-to-server UNSUBSCRIBE. */
    AZERO_IOT_MQTT_PINGREQ,           /**< Client-to-server PINGREQ. */
    AZERO_IOT_MQTT_DISCONNECT         /**< Client-to-server DISCONNECT. */
} azero_iot_mqtt_operation_type_e;

typedef enum azero_iot_mqtt_qos
{
    AZERO_IOT_MQTT_QOS_0 = 0, /**< Delivery at most once. */
    AZERO_IOT_MQTT_QOS_1 = 1, /**< Delivery at least once.*/
    AZERO_IOT_MQTT_QOS_2 = 2  /**< Delivery exactly once. */
} azero_iot_mqtt_qos_e;

#define AZERO_IOT_MQTT_MSG_INITIALIZER     { .pTopicName = NULL, .topicNameLength = 0, .pPayload = NULL, .payloadLength = 0}

typedef struct mqtt_payload_msg
{
    const char * pTopicName;          /**< @brief Topic name of PUBLISH. no longer than 23  */
    uint16_t topicNameLength;         /**< @brief Length of pTopicName. */

    const void * pPayload;            /**< @brief Payload of PUBLISH. */
    uint16_t payloadLength;             /**< @brief Length ofpPayload.*/
}mqtt_payload_msg_t;

typedef struct azero_iot_mqtt_info
{
    union
    {
        struct
        {
            
            const char* operationType;      /* string of @ref azero_iot_mqtt_operation_type_e */
            const char* operationResult;    /* `SUCCESS` or other error string */
        }operation; /* Operation complete callbacks. */

        mqtt_payload_msg_t subscrib; /* subscribed topics's callbacks. */
    } u;

}azero_iot_mqtt_info_t;

typedef void ( *mqttInfoCallback )( azero_iot_mqtt_info_t info );


typedef enum disconnect_reason
{
    AZERO_IOT_MQTT_DISCONNECT_CALLED,   /** @ref azero_iot_handler_release was invoked. */
    AZERO_IOT_MQTT_BAD_PACKET_RECEIVED, /** An invalid packet was received from the network. */
    AZERO_IOT_MQTT_KEEP_ALIVE_TIMEOUT,  /** Keep-alive response was not received. */
    AZERO_IOT_MQTT_NET_POLL_TIMEOUT     /** net layer pull data timeout**/
} disconnect_reason_t;

typedef void (*disconnectReasonCallback)( disconnect_reason_t reason);

typedef enum shadow_data_type
{
    AZERO_IOT_SHADOW_GET,       /** Get param saved in server. */
    AZERO_IOT_SHADOW_DELTA,     /** App ask to change device's state. */
    AZERO_IOT_SHADOW_DOCUMENT   /** Server update param success, send previous and current state. */
} shadow_data_type_t;

/**
 * @brief shadow's param callback
 */

typedef void ( *shadowParamCallback )( shadow_data_type_t type, const char *pDocument, size_t documentLength);

/**
 * result callback of @ref azero_iot_shadow_update
 */
typedef void ( *shadowUpdateResultCallback )( const char* result);

typedef struct azero_iot_callback
{
    mqttInfoCallback                operation_complete;
    mqttInfoCallback                subscrib_message;
    disconnectReasonCallback        disconnect;
    shadowParamCallback             param;
    shadowUpdateResultCallback      update_result;
} azero_iot_callback_t;

typedef struct azero_device_info
{
    char* phostname;        /* host name */
    uint16_t port;          /* port of host name */
    char* accountprefix;    /* created in azero iot web */
    char* pIdentifier;      /* device's clientId, must be unique, with azero server, combined accountprefix-sn */
    char* pThingName;       /* device's name, must be unique */
    bool  inputCertPath;    /* not support file read, make it false and input document directly, otherwise input path and make it true*/
    char* pRootCa;          /* trusted server root certificate*/
    char* pClientCert;      /* path of *cert.pem */
    char* pPrivateKey;      /* path of *private.key */
    char** pAttributeName;  /* such as color, brightness*/
    uint16_t attributeSize; /* size of attribute */
}azero_device_info_t;

typedef void * azero_iot_handler_t;

/**
 * @brief Create iot SDK handler.
 * @param [in] device_info, cbs must be global variable, SDK will use the info later
 * 
 * @return NULL if failed, other when success.
 */
SAI_API azero_iot_handler_t azero_iot_handler_create( azero_device_info_t* device_info, azero_iot_callback_t* cbs );

/**
 * @brief Destory iot SDK handler.
 * @param [in] handler
 */
SAI_API void azero_iot_handler_release( azero_iot_handler_t handler );

/**
 * @brief Start iot SDK.
 * @param [in] azero_iot_handler_t
 * @param [in] azero_iot_mqtt_operation_type_e, only use AZERO_IOT_MQTT_SUBSCRIBE and AZERO_IOT_MQTT_UNSUBSCRIBE.
 * @param [in] topics, a array of string
 * @param [in] qos, a array of azero_iot_mqtt_qos_e, using with same pos topic
 * @param [in] size of topics and qos
 * 
 * @return EXIT_SUCCESS when success, EXIT_FAILURE otherwise.
 */
SAI_API int azero_iot_mqtt_modify_subscriptions( azero_iot_handler_t handler,
                                                azero_iot_mqtt_operation_type_e operation, 
                                                const char ** topics, 
                                                const azero_iot_mqtt_qos_e qos[], 
                                                const int topicSize );

/**
 * @brief Publish msg async.
 * @param [in] azero_iot_mqtt_msg.
 *
 * @return EXIT_SUCCESS when publish success, EXIT_FAILURE if failed.
 */
SAI_API int  azero_iot_mqtt_publish_async( azero_iot_handler_t handler, 
                                            mqtt_payload_msg_t msg );

/**
 * @brief Publish msg sync.
 * @param [in] azero_iot_mqtt_msg.
 * @param [in] timeout microseconds, default 5000ms.
 *
 * @return EXIT_SUCCESS when publish success, EXIT_FAILURE if failed.
 */
SAI_API int  azero_iot_mqtt_publish_sync( azero_iot_handler_t handler, 
                                            mqtt_payload_msg_t msg,
                                            uint32_t timeoutMs );

/**
 * @brief Retrieves the latest state stored from server
 * 
 * * @return EXIT_SUCCESS when get success, EXIT_FAILURE if failed.
 */
SAI_API int azero_iot_shadow_get( azero_iot_handler_t handler);

/**
 * @brief Update shadow state async.
 * @brief Suggest reporting state with this method.
 * @param [in] pUpdateDocument, such as `{"color":"red"}`  or  `{"color":"red","power":"on"}`.
 * @param [in] resposeDesire, if is a response for desire.
 *             if device change the state after @ref changeDeviceStateDocumentCallback, 
 *             if true; otherwise is false.
 *
 * @return EXIT_SUCCESS when publish success, EXIT_FAILURE if failed.
 */
SAI_API int azero_iot_shadow_update( azero_iot_handler_t handler, const char* pUpdateDocument, bool resposeDesire );

/**
 * @brief Clear shadow document.
 * 
 */
SAI_API void azero_iot_shadow_clear( azero_iot_handler_t handler);

/**
 * @brief Parses a key in the pDeltaDocument.
 *
 * @param[in] pDeltaDocument The Shadow delta document to parse.
 * @param[in] deltaDocumentLength The length of `pDeltaDocument`.
 * @param[in] pDeltaKey The key in the delta document to find. Must be NULL-terminated.
 * @param[out] pDelta Set to the first character in the delta key.
 * @param[out] pDeltaLength The length of the delta key.
 *
 * @return `true` if the given delta key is found; `false` otherwise.
 */
SAI_API bool azero_iot_parser_value( const char * pDeltaDocument,
                                        size_t deltaDocumentLength,
                                        const char * pDeltaKey,
                                        const char ** pDelta,
                                        size_t * pDeltaLength );

/**
 * @brief Get the SDK version info.
 *
 * @return the SDK version info.
 */
SAI_API const char* azero_iot_get_sdk_version(void);

#if defined(__cplusplus)
}
#endif
#endif //SOUNDAI_AZERO_IOT_API_H

src/main.c
#include "azero_iot_sdk_api.h"
#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <pthread.h>

static azero_iot_handler_t handler = NULL;
static void startiot();
/* 设备支持的控制策略, 和Azero IoT平台控制策略一致 */
static char *state_name[] = {"brightness","lightMode"};
static int   state_size = 2;

#define ProductID        "speaker_sai"         /* 一类设备的product Id */
#define DeviceSN         "AE:DE:C2:61:3F:18"   /* 本设备的SN, 本类设备中每个设备唯一, 可以为MAC或其他唯一标识 */
#define AccountPrefix    "32141iz1awewfq"      /* 一类设备的标识, IoT平台注册产生 */

// 断开连接后的重连处理
static bool reconnecting = false;
static void * _reconnectThread(void* args)
{
    pthread_detach(pthread_self());
    reconnecting = true;
    azero_iot_handler_release(handler);
    handler = NULL;
    startiot();
    sleep(2);
    reconnecting = false;
    return NULL;
}

//断开连接的回调
static void disconnect_cb(disconnect_reason_t reason)
{
    printf("disconnect reason %d\n",reason);
    if(reason == AZERO_IOT_MQTT_KEEP_ALIVE_TIMEOUT | AZERO_IOT_MQTT_NET_POLL_TIMEOUT){
        if(false == reconnecting)
        {
            reconnecting = true;
            pthread_t ptid;
            pthread_create(&ptid, NULL, _reconnectThread,NULL);
        }
    }
}

static void deal_data_cb(const char *key, const char* value,size_t size, bool need_report)
{
    printf("get state= %s, value = %.*s,need report = %s\n",key,size,value,need_report?"true":"false");

    if(need_report){
        char *format = "{\"%s\":%.*s}";
        size_t length = strlen(format) + size + strlen(key);
        char data[length+1];
        sprintf(data,format,key,size,value);
        /* 端上状态上报, 如果是本地主动更改的状态, 最后一个参数传false, APP端发起的改变, 最后一个参数传true */
        azero_iot_shadow_update(handler,data,true);
    }
}

/*shadow服务接收到的消息回调*/
static void shadow_data_cb( shadow_data_type_t type, const char *pDocument, size_t documentLength)
{
    printf("shadow msg cb: %.*s\n",documentLength,pDocument);
    switch (type)
    {
    case AZERO_IOT_SHADOW_GET:
    case AZERO_IOT_SHADOW_DELTA:
        {
            bool stateFound = false, keyFound = false;
            const char * pState = NULL;
            size_t stateLength = 0;

            /* 查找json消息中是否有"state" key. */
            stateFound = azero_iot_parser_value( pDocument,
                                                    documentLength,
                                                    "state",
                                                    &pState,
                                                    &stateLength );
            if(!stateFound)
            {
                printf("pDocument is error\n");
                return;
            }
            /* 查找是否有已经定义好的策略 */
            printf("state = %.*s\n",stateLength,pState);
            for(int i=0; i<state_size; i++){
                
                const char * pData = NULL;
                size_t dataLength = 0;
                const char * keyName = state_name[i];
                keyFound = azero_iot_parser_value( pState,
                                                    stateLength,
                                                    keyName,
                                                    &pData,
                                                    &dataLength );
                printf(" find key = %s, result = %d\n",state_name[i],keyFound);
                if(keyFound)
                {
                    /* 对于get到的消息, 需要设备改变状态至相应的状态, 和IoT平台最新消息保持一致,
                    一般用于设备开机或者断网重连, 对APP端做出的改变进行同步;
                    delta 消息为APP下发控制指令后收到的消息, 需要将设备改变至相应状态, 同时将改变后状态上报IoT平台,
                    平台更新最新状态.
                    */
                    deal_data_cb(keyName,pData,dataLength,type==AZERO_IOT_SHADOW_DELTA);
                }
            }

            break;
        }
    case AZERO_IOT_SHADOW_DOCUMENT:
        {
            /* update后云端当前状态和上一个状态的通知, 可以依次查看是否上报错误 */
            break;
        }
    }
}

/* shadow 内容update结果回调, 成功或者失败*/
static void update_result_cb(const char* result)
{
    printf("update result is %s\n",result);
}

static void startiot()
{
    while(handler == NULL){
        /* global val, SDK will use this info later */
        static azero_iot_callback_t cbs = { .disconnect = disconnect_cb,\
                                            .param = shadow_data_cb,\
                                            .update_result = update_result_cb};
        
        static azero_device_info_t device_info = { .phostname = AccountPrefix".iot-dev.bj.soundai.cn",\ 
                                                    /* IoT 平台产生域名 */
                                                    .port = 8883,\
                                                    .accountprefix = AccountPrefix,\
                                                    .pIdentifier = AccountPrefix"-"DeviceSN,\
                                                    /* pIdentifier 每个设备唯一, 前缀必须为AccountPrefix */
                                                    .pThingName = ProductID""DeviceSN,\
                                                    /* pThingName 每个设备唯一, 必须为ProductIDDeviceSN拼接 */
                                                    /* IoT 平台产生以下key */
                                                    .inputCertPath = true,\
                                                    .pRootCa = "1323214-ca.crt",\
                                                    .pClientCert = "1323214-cert.pem",\
                                                    .pPrivateKey = "1323214-private.key.pem"};
        device_info.pAttributeName = state_name;
        device_info.attributeSize = state_size;
        for(int i=0;i<state_size;i++)
        {
            printf("pAttributeName[%d] = %s\n",i,device_info.pAttributeName[i]);
        }

        handler = azero_iot_handler_create(&device_info,&cbs);
        sleep(2);        
    }
}

int main()
{
    startiot();

    azero_iot_shadow_get(handler);
    while(1){
        sleep(rand()%50+5);
        /* azero_iot_shadow_update(handler,"{\"brightness\":33}",false);*/
    };
    /*azero_iot_shadow_clear(handler);
    清空IoT 平台所有状态, 用于恢复出厂设置或者重置
    */
    /* 正常退出, 释放资源 */
    azero_iot_handler_release(handler);
    handler = NULL;
}
